source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,915 @@
+    if (inst.opcode() == SpvOpPhi &&
+        (inst.result_id() != induction->result_id() &&
+         ir_context_->get_def_use_mgr()->NumUsers(&inst) != 0)) {
+      return false;
+    }
+  }
+
+  // Can only unroll inner loops.
+  if (!loop->AreAllChildrenMarkedForRemoval()) {
+    return false;
+  }
+
+  for (uint32_t block_id : loop->GetBlocks()) {
+    opt::analysis::DefUseManager* def_use_manager =
+        ir_context_->get_def_use_mgr();
+
+    ir::BasicBlock& bb = *ir_context_->cfg()->block(block_id);
+    // For every instruction in the block.
+    for (ir::Instruction& inst : bb) {
+      if (inst.result_id() == 0) continue;
+
+      auto is_used_outside_loop = [loop,
+                                   def_use_manager](ir::Instruction* user) {
+
+        if (!loop->IsInsideLoop(user)) {
+          // Some optimization passes have been seen to leave dead phis in the
+          // IR so we check that if a phi is used outside of the loop that the
+          // user
+          // is not dead.
+          if (!(user->opcode() == SpvOpPhi &&
+                def_use_manager->NumUsers(user) == 0))
+            return false;
+        }
+        return true;
+      };
+
+      if (!def_use_manager->WhileEachUser(&inst, is_used_outside_loop)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool LoopUtils::PartiallyUnroll(ir::Loop* loop, size_t factor) {
+  if (factor == 1 || !CanPerformUnroll(loop)) return false;
+
+  // Create the unroller utility.
+  LoopUnrollerUtilsImpl unroller{ir_context_, &function_};
+  unroller.Init(loop);
+
+  // If the unrolling factor is larger than or the same size as the loop just
+  // fully unroll the loop.
+  if (factor >= unroller.GetLoopIterationCount()) {
+    unroller.FullyUnroll(loop);
+    return true;
+  }
+
+  // If the loop unrolling factor is an residual number of iterations we need to
+  // let run the loop for the residual part then let it branch into the unrolled
+  // remaining part. We add add one when calucating the remainder to take into
+  // account the one iteration already in the loop.
+  if (unroller.GetLoopIterationCount() % factor != 0) {
+    unroller.PartiallyUnrollResidualFactor(loop, factor);
+  } else {
+    unroller.PartiallyUnroll(loop, factor);
+  }
+
+  return true;
+}
+
+bool LoopUtils::FullyUnroll(ir::Loop* loop) {
+  if (!CanPerformUnroll(loop)) return false;
+
+  LoopUnrollerUtilsImpl unroller{ir_context_, &function_};
+
+  unroller.Init(loop);
+  unroller.FullyUnroll(loop);
+
+  return true;
+}
+
+void LoopUtils::Finalize() {
+  // Clean up the loop descriptor to preserve the analysis.
+  loop_descriptor_.PostModificationCleanup();
+}
+
+/*
+ *
+ * Begin Pass.
+ *
+ */
+
+Pass::Status LoopUnroller::Process(ir::IRContext* c) {
+  context_ = c;
+  bool changed = false;
+  for (ir::Function& f : *c->module()) {
+    LoopUtils loop_utils{&f, c};
+
+    for (ir::Loop& loop : loop_utils.GetLoopDescriptor()) {
```

Comment:
Do you do any ordering with regards to loop nesting? If you unroll an outer loop won't you produce a large number of new loops which you'll then need to unroll, or does this pass not support nested loops?

Reply:
It doesn't support unrolling inner most loops, if you unroll nested loops you have to go inside out. The CanPerfomUnroll function checks that the loop has no children

Reply:
So does this mean that if the loop descriptor iterator gives you the inner most loops first, it will unroll them, but if you get them in outer most first order unrolling won't happen. Am I understanding correctly?

Reply:
Yes, and it does give them in that order